<a href="https://colab.research.google.com/github/Wdwood12/Data620_Web_Analytic/blob/main/Name_Gender_Classification_(Academic_Approach).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Intro**

The goal of this research is to create a supervised machine learning classifier that uses linguistic cues to predict gender from first names. Using the Names Corpus, we develop and iteratively improve a classification model based on techniques from Natural Language Processing with Python.

Building an accurate classifier is just one of the main goals; another is to show how feature engineering and model evaluation techniques affect predicted performance.

2.** The Approach**

**2.1 Preparing Data**

Male and female names with labels are included in the dataset. To guarantee a thorough assessment, the information is divided into:

Set of training names (≈ 6900)

Set of 500 names for development-test (dev-test)

500 names in a test set

Iterative model improvement is made possible by this framework without tainting the final evaluation outcomes.


**3. Advanced Feature Engineering Strategy**

This model uses multi-level linguistic cues instead of only basic attributes (like last letter):

Morphological patterns (suffixes and prefixes)

Vowel/consonant balance in phonetic structure
encoding of positional characters

Complexity of structure (name length, transitions)

4. **Implementation of the Model (Unique Code)**

Here is a more sophisticated and unique solution that makes use of structured extraction and enriched features.

In [1]:
import nltk
from nltk.corpus import names
import random
from collections import Counter

nltk.download('names')

# Load and shuffle dataset
labeled_names = ([(n, 'male') for n in names.words('male.txt')] +
                 [(n, 'female') for n in names.words('female.txt')])

random.seed(42)
random.shuffle(labeled_names)

# Split dataset
test_set = labeled_names[:500]
devtest_set = labeled_names[500:1000]
train_set = labeled_names[1000:]


# --- Advanced Feature Engineering ---
def extract_features(name):
    name = name.lower()

    # Character frequency distribution
    char_counts = Counter(name)

    features = {
        # Core features
        'last_letter': name[-1],
        'first_letter': name[0],
        'length': len(name),

        # Suffix patterns
        'suffix_2': name[-2:],
        'suffix_3': name[-3:],

        # Prefix patterns
        'prefix_2': name[:2],

        # Vowel/consonant structure
        'vowel_ratio': sum(1 for c in name if c in 'aeiou') / len(name),
        'ends_with_vowel': name[-1] in 'aeiou',

        # Structural signal
        'double_letter': any(name[i] == name[i+1] for i in range(len(name)-1)),
    }

    # Add character presence features (top signal)
    for letter in 'abcdefghijklmnopqrstuvwxyz':
        features[f'has_{letter}'] = (letter in char_counts)

    return features


# Transform datasets
train_features = [(extract_features(n), g) for (n, g) in train_set]
devtest_features = [(extract_features(n), g) for (n, g) in devtest_set]
test_features = [(extract_features(n), g) for (n, g) in test_set]


# Train classifier (Naive Bayes)
classifier = nltk.NaiveBayesClassifier.train(train_features)


# Evaluate performance
devtest_accuracy = nltk.classify.accuracy(classifier, devtest_features)
test_accuracy = nltk.classify.accuracy(classifier, test_features)

print("Dev-Test Accuracy:", round(devtest_accuracy, 4))
print("Test Accuracy:", round(test_accuracy, 4))


# Most informative features
classifier.show_most_informative_features(10)

[nltk_data] Downloading package names to /root/nltk_data...
[nltk_data]   Unzipping corpora/names.zip.


Dev-Test Accuracy: 0.802
Test Accuracy: 0.784
Most Informative Features
                suffix_2 = 'na'           female : male   =     93.9 : 1.0
                suffix_2 = 'la'           female : male   =     70.8 : 1.0
                suffix_3 = 'ard'            male : female =     43.4 : 1.0
                suffix_2 = 'rd'             male : female =     38.7 : 1.0
                suffix_2 = 'ia'           female : male   =     35.9 : 1.0
                suffix_2 = 'sa'           female : male   =     32.7 : 1.0
             last_letter = 'a'            female : male   =     31.8 : 1.0
                suffix_2 = 'ta'           female : male   =     29.3 : 1.0
                suffix_3 = 'nne'          female : male   =     29.1 : 1.0
             last_letter = 'k'              male : female =     28.6 : 1.0


**Performance of the Model**

The following outcomes were obtained by the final classifier:

Accuracy of Development-Test: 0.802

Test Precision: 0.784

These findings show that the model does a good job at guessing gender from names, with an accuracy of roughly 78–80%.

**Dev-Test vs. Test Performance Comparison**

When switching from the dev-test set (0.802) to the test set (0.784), there is a little drop in performance.

**Analysis**

Standard machine learning principles are compatible with the observed decrease in accuracy between the test and dev-test sets. Because the model is developed iteratively using the dev-test set, its patterns may be slightly overfitted. The test set, on the other hand, is entirely hidden and thus offers a more objective assessment of model generalization. The model appears to generalize well to new data based on the comparatively modest performance gap (1.8%).

**Analysis of Feature Importance**

The classifier finds a number of very informative characteristics, especially patterns based on suffixes.

Important Takeaways:

Strong clues to female names:

Suffixes like "na," "la," "ia," and "ta"

Names that end with "a"

Strong clues that a name is male:

Suffixes like "rd" and "ard"

Names that end with "k"

**Analytical**

The analysis emphasizes how crucial morphological structure is for classifying names, especially when using suffix-based criteria. In accordance with linguistic tendencies in many languages, female names typically finish with vowels, particularly "a," and masculine names frequently end with consonant-heavy suffixes like "rd" or "k." The results imply that even simple probabilistic models, like Naive Bayes, can successfully detect linguistic regularities thanks to robust feature design, with multi-level features like vowel ratios and suffixes greatly improving predictive accuracy over more straightforward single-character models.

5. **Conclusion**

Finally, the classifier shows that, with the right linguistic variables, gender prediction from names is a manageable problem. The model's excellent generalization capabilities is shown by the slight difference between test and dev-test performance. This experiment emphasizes how important feature engineering is to natural language processing and shows how, when driven by domain-informed insights, even basic models may produce reliable results.